### ISSUE TITLE
JSPanel 4 alpha 83

### ISSUE THREAD

**Issue by RobertoMalatesta** [Flyer53/jsPanel3#63]
Hi Stefan,

I'm just having a look at Alpha 83, and I noticed that (on mobile) the collapse/minimize/close buttons sometimes do not react to first tap. It could be related to the 300ms delay, maybe.

<hr>
Another thing that can be useful to you:
`//@ts-check `
is an option that, if put inside JS files allows Typescript compiler to give hints in JS code without the need of converting it in another language.
It can catch suspicious things like these:

```
source/jspanel.js(1467,57): error TS2339: Property 'forEach' does not exist on type 'NodeListOf<Element>'.
source/jspanel.js(1520,13): error TS2322: Type '90' is not assignable to type 'string'.
source/jspanel.js(1531,21): error TS2322: Type 'NodeListOf<HTMLIFrameElement>' is not assignable to type 'any[]'.
source/jspanel.js(1836,53): error TS2339: Property 'front' does not exist on type 'HTMLElement'.
source/jspanel.js(2041,29): error TS2304: Cannot find name '$'.
...
source/jspanel.js(2293,17): error TS2322: Type '"filled"' is not assignable to type 'boolean'.
source/jspanel.js(2296,17): error TS2322: Type '"filledlight"' is not assignable to type 'boolean'.
source/jspanel.js(2299,17): error TS2322: Type '""' is not assignable to type 'boolean'.
source/jspanel.js(2302,13): error TS2322: Type 'any[]' is not assignable to type 'boolean'.
source/jspanel.js(2305,29): error TS2339: Property 'match' does not exist on type 'boolean'.
```

These days I'm using this functionality a lot and I almost write no TS code at all but assisted-JS, since it's more fun and productive.

For a brief intro to the functionality see:
[https://github.com/Microsoft/TypeScript/wiki/Type-Checking-JavaScript-Files](https://github.com/Microsoft/TypeScript/wiki/Type-Checking-JavaScript-Files)

Feel free to ask me if you want any help I can provide.

<hr>

Then, some questions still on Rel.4:

- How much the API of ver. 4 alpha can be considered stable?
- What functionalities are still missing compared to 3.x and 2.x ?
- The idea of not depending on other frameworks is a damn winning one, since it can be easily integrated with *any* framework
- Are you planning to create a separate github repo for it?

Zum nächsten mal,

--R

**Comment by Flyer53** [comment]
@RobertoMalatesta 
Hey Rob, first of all thanks again for your input :)

**300ms**
I personally don't think it's related to the 300ms issue: On a touch device jsPanel v4 controls use the **pointerup** event if supported, otherwise the **touchend** event. Those events are fired instantly as far as I know, without any delay. So it must be something else I think.
This "no reaction to first tab" thing is something I observe on a lot of websites.

**TS**
Sounds interesting, I'll take a look at this feature.
Btw: A NodeList.prototype.forEach() polyfill is included in jsPanel v4

**v4 alpha API etc.**
+ I think the API is stable. In fact I plan to make it a Beta pretty soon and I don't have any API changes in mind.
+ As far as I see it v4 is complete and there are no things left from v3 or v2 I want to have in v4. On purpose I did not implement an **option.contentIframe** in v4 since it's absolutely no problem to add an iframe with option.content.
+ I tried hard to do it without dependencies ... and don't want to introduce one again 😏 
+ The repo: That's an issue. I would like to have everything in one repo but can't make up my mind whether it would be good to add all v3 releases to the original **jsPanel** repo and than continue with v4 also in the original jsPanel repo. I somehow don't like the idea of a jsPanel4 repo.

Happy coding,
Stefan 😄 

PS: If you see something missing in v4 don't hesitate to tell me!

**Comment by RobertoMalatesta** [comment]
@Flyer53 
Hi Stefan,
just got one implementing a horizontal toolbar:

   
```
let tb = 
<div class="jsPanel-content jsPanel-content-nofooter horizontal-toolbar" 
    style="border-top: 1px solid rgb(0, 0, 0); width: 100%; height: 50px; overflow: hidden;">
      <i class="fa fa-user"></i><i class="fa fa-cog">
        </i><i class="fa fa-calendar">
          </i><i class="fa fa-floppy-o"></i><i class="fa fa-camera"></i>
          <i class="fa fa-picture-o"></i><i class="fa fa-print"></i>
          <i class="fa fa-map-marker"></i></div>

    jsPanel.create({
      theme: 'darkslateblue filled',
      contentSize: { width: 475, height: 50 },
      headerRemove: true,
      resizeit: false,
      dragit: { handles: 'div.jsPanel-content' },
      content: tb,
      callback: function (panel) {
        // added close icon handler
        $('span.fa-remove', panel.content).on('click', function () { panel.close(); });
        panel.content.css('cursor', 'move');
      }
    }).content.addClass('horizontal-toolbar');


```


The code is from your example at 
[http://jspanel.de/api/#examples/toolbars](http://jspanel.de/api/#examples/toolbars)
and it works perfectly under 3.x.
On v4a83 if I try to move the bar the cursor gets caught from the content and the bars start following it up to the next reload.

even removing the *.content.addClass('horizontal-toolbar');* that gives error does not fix it.
Has something changed in the dragit?

--R

**Comment by RobertoMalatesta** [comment]
*300 ms:* 
> This "no reaction to first tab" thing is something I observe on a lot of websites.

hm, I read something a couple of months ago and tried the 2-three main libs that offer mobile dragging: all of them had issues.
I will investigate more.

*ts*: 
>  A NodeList.prototype.forEach() polyfill is included in jsPanel v4

Btw if you set in the tsconfig.json something like:
 "lib": [
      "dom",
      "es6"
    ],
that gives similar polyfills for free.
(being able to still target es3 and --outFile emits a single file ready to be compressed/obfuscated)

*v4*:

>  On purpose I did not implement an option.contentIframe in v4

No need indeed. I still have to reproduce in a test that bug I mentioned in the other open case but I'm not using IFrames now.

> I tried hard to do it without dependencies ... and don't want to introduce one again

I totally agree on this. By adding this and that one ends up with a... a... a  *Jamaica*! :laughing: 
That's why I suppose you'll end up giving a clear cut between v3 and v4 and create a new repo.
A lot of work tho.

> PS: If you see something missing in v4 don't hesitate to tell me!

It would be interesting experimenting with a pure *flexbox-based* panel system.
Besides that, I find jsPanel rather complete.

> Happy coding,

Beer time, now 
Ideally sharing a couple of Lager with you, my friend! :beers:

--Rob

**Comment by Flyer53** [comment]
@RobertoMalatesta 
**horizontal toolbar**
Look at the toolbar HTML. You included a div.jsPanel-content resulting in 2 divs with the class jsPanel-content. That's the problem. Just remove the div from the toolbar code and it should work.
```javascript
var tb = '<i class="fa fa-user fa-2x"></i>' +
'<i class="fa fa-cog fa-2x"></i>' +
'<i class="fa fa-calendar fa-2x"></i>' +
'<i class="fa fa-camera fa-2x"></i>' +
'<i class="fa fa-print fa-2x"></i>' +
'<i class="fa fa-map-marker fa-2x"></i>';

jsPanel.create({
    theme: 'darkslateblue filled',
    contentSize: {width: 475, height: 45},
    headerRemove: true,
    resizeit: false,
    dragit: {handles: 'div.jsPanel-content'},
    content: tb,
    callback: function(){
        this.content.querySelectorAll('i').forEach(function(item){
            item.style.margin = '10px';
        });
    }
});
```

**flexbox**
Header and footer sections in v4 already use flexbox. The rest should be up to the developer I think.
Or how do you mean that?

Sharing a beer sounds good ... but I fear we're not just round the corner, are we?
And please don't mention Jamaica again, unless you're talking about Cocktails 😏

**Comment by RobertoMalatesta** [comment]
> horizontal toolbar

My bad, thank you.
I noticed that the object returned by jsPanel.create() still has _.content_ but no _.content.addClass()_ method.
Is it on purpose / I'm missing something ?
I bet on the second.

**Comment by Flyer53** [comment]
No problem at all... I have yet to add some of those examples to the v4 docs.

In v3 the content property is a jQuery object and the .addClass() method you see is the jQuery method.

In v4 the content property is the element itself and you can use native js like:
```javascript
panel.content.classList.add("myPanel"); // e.g.
```
https://developer.mozilla.org/en-US/docs/Web/API/Element/classList
No need to add a jsPanel specific method I think.

**Comment by Flyer53** [comment]
Almost missed that in v4 there are the global jsPanel methods
+ jsPanel.setClass() - http://alpha.jspanel.de/docs/#global/setClass
+ jsPanel.removeClass() - http://alpha.jspanel.de/docs/#global/remClass
+ jsPanel.setStyle() - http://alpha.jspanel.de/docs/#global/setStyle

that to a certain degree mimic the corresponding jQuery methods.

**Comment by Flyer53** [comment]
Just made a small change  and uploaded [v4-alpha.84](http://alpha.jspanel.de.localhost/downloads/jspanel-4-alpha.zip)
Among a few minor corrections I added css *touch-action: none* to the controlbar. Might improve the "no-action-on-first-click" issue ....

**Comment by RobertoMalatesta** [comment]
Tried it, but I still get those event losses.
Sometimes when maximized the restore to size icon stays grey after being pressed but no resize takes place.

**Comment by RobertoMalatesta** [comment]
touch-action: none was present in 83 as well.
I tried with a  touch-action: manipulation; in .jsPanel-controlbar since 
[https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action](https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action) says :  
touch-action [:manipulation] is also often used to completely disable the delay of click events caused by support for the double-tap to zoom gesture.

**Comment by RobertoMalatesta** [comment]
```
.jsPanel-btn {
  touch-action: manipulation !important;
}
```

tried this since this should remove the delay on the div containing the button controls.
All times that I press a button and I don't get the desired feedback the button changes background as if it had gotten the event.

**Comment by Flyer53** [comment]
I prefer *touch-action: none* over *manipulation* because the browser shall not do any panning and zooming gestures in regard of the controls.

v4 controls on mobiles either listen for a *pointerup* or *touchend* event. Those events fire instantly without any kind of delay. See https://www.sitepoint.com/5-ways-prevent-300ms-click-delay-mobile-devices/ heading 4 and 5. So this is not the problem.

I guess it's much more simple: When you use a relatively large finger to work the relatively small controls it will sometimes happen that the touched surface outside the control is larger than the touched surface inside the control and thus firing an event for the element next to the control but not the control itself while still showing the css hover effect. I don't know whether that's the way a browser works, but it sounds reasonable to me. I came to this conclusion because on my two touch devices I could never ever reproduce this issue using a pen but can reproduce it using a finger. Did you ever try using a pen?

**Comment by RobertoMalatesta** [comment]
> Did you ever try using a pen?
Yep. I have a couple of small felt-tip tablet pointers.
A way of debugging is having some div change color when he button is clicked and write the event.
Not a prio but a nagging ish. I will do it.

**Comment by Flyer53** [comment]
@RobertoMalatesta 
Just for info: Access to http://www.exedraos.com/ is blocked by my anti-virus due to a found virus **Script.Packed.Agent.F@susp**

**Comment by RobertoMalatesta** [comment]
Thank you Stefan,
False positive.
Just encrypted the portal the hard way with a nice project here on github called javascript-obfuscator.
Glad you spot this problem with Windows Antivirus: it's something I tend to ignore since my machines are all Linux since 2001.
And when I do have to use Windows for work, it's as a vmware/virtual box instance.
You probably already know it, but, just in case:
Pros of virtualized Windows are many: 
- multiple versions of Windows, 
- no fear of viruses or trojans, sinceI don't keep any valuable info on those vms, 
- easy-peasy to backup: it's a file copy
- if an installed instance goes unstable you simply pick up a former backup

exedraos.com should now be back as original.
Can you tell me if you still have the error after a cache clean?

Thanks again,

--Rob

**Comment by Flyer53** [comment]
@RobertoMalatesta 
👍 on exedraos.com

**Comment by RobertoMalatesta** [comment]
I'll have to take the time to setup some Virtual Windows Box with all the browsers and all the main AV products.

Today the surface of attack of browser is increasing and Antivirus companies are clueless and scared as hell.
AV are viruses in themselves, as they create more problems than solving real ones, imho.

Msg me directly on twitter at exedraos if you'd like to do some tests on that page source and discuss real  browsers vulnerabilities as well.

Have a good week-end !

--R

PS: it seems to me that adding a:

```
.jsPanel-btn {
  touch-action: manipulation;  // maybe !important can be left out for tidyness
}
```

reduces the issues about responsivity in jsPanel.

**Comment by Flyer53** [comment]
Thanks for the offer Rob 😄 
I stick to the same AV for years now and so far I didn't have any problems. Hopefully I was not only lucky ...

Mmmmh, might need to play around with the css some more ... but for now it's weekend, you're right on that 😏 

Just uploaded [v4 alpha.85](http://alpha.jspanel.de/downloads/jspanel-4-alpha.zip) fixing a bug in the extensions tooltip and contextmenu

Have a nice one too,
Stefan

**Comment by Flyer53** [comment]
@RobertoMalatesta 
Hi Rob,
I just released jsPanel v4.0.0-beta.1 in its own GitHub repo.
So if an issue comes up please file it in the new repo.
Happy coding, Stefan 😄
